### PR TITLE
Add isolated cpu parsing

### PR DESCRIPTION
**Pull request by DavidVentura**
Hi
This PR adds support for parsing `/sys/devices/system/cpu/isolated`
Prompted by [this PR](https://github.com/prometheus/node_exporter/pull/2251) for node_exporter

I am not sure if this should be a `uint16` or not.

### CODE REVIEW

File: sysfs/system_cpu.go [hunk 1]
```diff
@@ -237,3 +240,47 @@ func parseCpufreqCpuinfo(cpuPath string) (*SystemCPUCpufreqStats, error) {
...
+func (fs FS) IsolatedCPUs() ([]uint16, error) {
+	isolcpus, err := ioutil.ReadFile(fs.sys.Path("devices/system/cpu/isolated"))
+	if err != nil {
+		return nil, fmt.Errorf("failed to read isolcpus from sysfs: %w", err)
+	}
+
+	return parseIsolCpus(isolcpus)
+}
+
+func parseIsolCpus(data []byte) ([]uint16, error) {
```
Comment (discordianfish) [inline sysfs/system_cpu.go:253]: nit: lets call it `parseIsolatedCPUs`

File: sysfs/system_cpu.go [hunk 2]
```diff
@@ -237,3 +240,47 @@ func parseCpufreqCpuinfo(cpuPath string) (*SystemCPUCpufreqStats, error) {
...
+	isolcpus_str := strings.TrimRight(string(data), "\n")
+
+	var isolcpus_int = []uint16{}
+
+	for _, cpu := range strings.Split(isolcpus_str, ",") {
+		if cpu == "" {
+			continue
+		}
+		if strings.Contains(cpu, "-") {
+			ranges := strings.Split(cpu, "-")
```
Comment (discordianfish) [inline sysfs/system_cpu.go:263]: Check if len(ranges) == 2, otherwise accessing ranges[1] might panic.

File: sysfs/system_cpu.go [hunk 3]
```diff
@@ -237,3 +240,47 @@ func parseCpufreqCpuinfo(cpuPath string) (*SystemCPUCpufreqStats, error) {
...
+	isolcpus, err := ioutil.ReadFile(fs.sys.Path("devices/system/cpu/isolated"))
+	if err != nil {
+		return nil, fmt.Errorf("failed to read isolcpus from sysfs: %w", err)
+	}
+
+	return parseIsolCpus(isolcpus)
+}
+
+func parseIsolCpus(data []byte) ([]uint16, error) {
+	isolcpus_str := strings.TrimRight(string(data), "\n")
```
Comment (discordianfish) [inline sysfs/system_cpu.go:254]: idiomatic golang uses camel case, same with isolcpus_int.

File: sysfs/system_cpu.go [hunk 4]
```diff
@@ -237,3 +240,47 @@ func parseCpufreqCpuinfo(cpuPath string) (*SystemCPUCpufreqStats, error) {
...
+
+	var isolcpus_int = []uint16{}
+
+	for _, cpu := range strings.Split(isolcpus_str, ",") {
```
Comment (discordianfish) [inline sysfs/system_cpu.go:258]: I'd just do `range strings.Split(strings.TrimRight(string(data), "\n")`, no need for isolcpus_str

File: sysfs/system_cpu.go [hunk 5]
```diff
@@ -237,3 +240,47 @@ func parseCpufreqCpuinfo(cpuPath string) (*SystemCPUCpufreqStats, error) {
...
+				return nil, err
+			}
+
+			for i := startRange; i <= endRange; i++ {
+				isolcpus_int = append(isolcpus_int, uint16(i))
+			}
+			continue
+		}
+
+		_cpu, err := strconv.Atoi(cpu)
```
Comment (discordianfish) [inline sysfs/system_cpu.go:279]: _ as prefix is also uncommon, I'd just call it cpuN or something

File: sysfs/system_cpu_test.go [hunk 6]
```diff
@@ -140,3 +140,43 @@ func TestSystemCpufreq(t *testing.T) {
...
+	}{
+		{[]byte(""), []uint16{}, nil},
+		{[]byte("1\n"), []uint16{1}, nil},
+		{[]byte("1"), []uint16{1}, nil},
+		{[]byte("1,2"), []uint16{1, 2}, nil},
+		{[]byte("1-2"), []uint16{1, 2}, nil},
+		{[]byte("1-3"), []uint16{1, 2, 3}, nil},
+		{[]byte("1,2-4"), []uint16{1, 2, 3, 4}, nil},
+		{[]byte("1,3-4"), []uint16{1, 3, 4}, nil},
+		{[]byte("1,3-4,7,20-21"), []uint16{1, 3, 4, 7, 20, 21}, nil},
```
Comment (discordianfish) [inline sysfs/system_cpu_test.go:158]: Would be nice to add a test cases that returns an error. For example "1,2-,3" which as of today would panic.

**Comment by discordianfish** [comment]
Beside that, looks good - especially for a first time go contribution 🥳

File: sysfs/system_cpu.go (hunk 2 above)
Comment (DavidVentura) [reply sysfs/system_cpu.go:263]: i thought it'd panic as well, but the split still has 2 items in the slice (`'1-' -> ["1", ""]`)

File: sysfs/system_cpu.go [hunk 7]
```diff
@@ -237,3 +240,49 @@ func parseCpufreqCpuinfo(cpuPath string) (*SystemCPUCpufreqStats, error) {
...
+func (fs FS) IsolatedCPUs() ([]uint16, error) {
+	isolcpus, err := ioutil.ReadFile(fs.sys.Path("devices/system/cpu/isolated"))
+	if err != nil {
+		return nil, fmt.Errorf("failed to read isolcpus from sysfs: %w", err)
+	}
+
+	return parseIsolCPUs(isolcpus)
+}
+
+func parseIsolCPUs(data []byte) ([]uint16, error) {
```
Comment (discordianfish) [inline sysfs/system_cpu.go:253]: nit: Can you rename this to parseIsolatedCPUs for consistency? Thanks!

**Review by discordianfish — APPROVED** [review]
Beside that last nit, LGTM. Thanks!

**Review by SuperQ — APPROVED** [review]

**Pull request merged**